[Merged by Bors] - feat(Geometry/Convex/Cone): tensor products of pointed cones

bjornsolheim opened the pull request:
Define the minimal and maximal tensor product of pointed cones.
Prove that the minimal tensor product is less than or equal to the maximal tensor product.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

- [x] depends on: #25292


[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

bjornsolheim committed:
feat(Geometry/Convex/Cone): add conical combinations and conical hull

bjornsolheim committed:
feat(Geometry/Convex/Cone): add conical combinations and conical hull

bjornsolheim committed:
Add tensor product of pointed cones

- Added TensorProduct.lean with minimal and maximal tensor products
- minTensorProduct: conical hull of elementary tensors
- maxTensorProduct: dual cone characterization
- Proved minTensorProduct ≤ maxTensorProduct

bjornsolheim committed:
Added newline at the end

bjornsolheim committed:
Minor fix to adjust for changes in ConicalHull

bjornsolheim committed:
Renamed the file and made requested changes

bjornsolheim committed:
Minor fix to accommodate changes in ConicalCombination

bjornsolheim committed:
Updated the file reference to accommodate the name change

bjornsolheim committed:
removed ConicalHull.lean

bjornsolheim committed:
Removed ConicalCombination.lean

bjornsolheim committed:
Changed proof to use Submodule.span_le, removed to do

bjornsolheim committed:
Removed ConicalCombination.lean

YaelDillies commented on Mathlib/Geometry/Convex/Cone/TensorProduct.lean:
```diff
@@ -0,0 +1,122 @@
+    have h_pos_z₁ : 0 ≤ dualDistrib K G H (φ₁ ⊗ₜ[K] φ₂) z₁ := hz₁ φ₁ φ₂ hφ₁ hφ₂
+    have h_pos_z₂ : 0 ≤ dualDistrib K G H (φ₁ ⊗ₜ[K] φ₂) z₂ := hz₂ φ₁ φ₂ hφ₁ hφ₂
+    simp [map_add]
+    positivity
```
Can you write this as the dual of the min tensor product of the dual?

YaelDillies commented on Mathlib/Geometry/Convex/Cone/TensorProduct.lean:
```diff
@@ -0,0 +1,122 @@
+/-- The set of elementary tensor products from two cones: all tensors of the form x ⊗ₜ y
+where x ∈ C₁ and y ∈ C₂. This forms the generating set for the minimal tensor product. -/
+def elementaryTensors (C₁ : PointedCone K G) (C₂ : PointedCone K H) : Set (G ⊗[K] H) :=
+  {z | ∃ x y, x ∈ C₁ ∧ y ∈ C₂ ∧ z = x ⊗ₜ[K] y}
```
I don't think we want this definition. Can you inline it below?

bjornsolheim committed:
Implemented requested changes and other minor fixes

bjornsolheim committed:
Removed redundant namespace prefixes

YaelDillies commented on Mathlib/Geometry/Convex/Cone/TensorProduct.lean:
```diff
@@ -0,0 +1,114 @@
+## References
+
+[Aubrun et al. "Entangleability of cones"] Geom. Funct. Anal. Vol. 31 (2021) 181–205
+available as https://arxiv.org/abs/1911.09663
```
Can you turn this into a proper reference?

bjornsolheim committed:
Added proper reference

bjornsolheim committed:
switched to '*' for unordered lists

bjornsolheim committed:
Corrected "ticks" to "backticks" so that it creates the proper hyperlinks

YaelDillies commented on Mathlib/Geometry/Convex/Cone/TensorProduct.lean:
```diff
@@ -0,0 +1,113 @@
+/-- The minimal tensor product of two cones is given by all conical combinations of elementary
+tensor products x ⊗ₜ y of cone elements x and y. -/
+noncomputable def minTensorProduct (C₁ : PointedCone K G) (C₂ : PointedCone K H)
+    : PointedCone K (G ⊗[K] H) :=
```
```suggestion
noncomputable def minTensorProduct (C₁ : PointedCone K G) (C₂ : PointedCone K H) :
    PointedCone K (G ⊗[K] H) :=
```
Same below

bjornsolheim committed:
Corrected, lines should not start with ":"

bjornsolheim committed:
Reduced assumptions to CommRing

bjornsolheim committed:
minor formatting changes

bjornsolheim committed:
Changed variable name to R

bjornsolheim committed:
Removed unnecessary namespace prefix.
Removed unnecessary parenthesis.
Applied set coercion instead of .carrier.
Implemented image2 instead of set constructor notation.
Updated broken proofs.

YaelDillies commented on Mathlib/Geometry/Convex/Cone/TensorProduct.lean:
```diff
@@ -97,16 +89,17 @@ theorem tmul_mem_maxTensorProduct {x y} {C₁ : PointedCone R G} {C₂ : Pointed
 /-- The minimal tensor product is less than or equal to the maximal tensor product. -/
 theorem minTensorProduct_le_maxTensorProduct (C₁ : PointedCone R G) (C₂ : PointedCone R H) :
     minTensorProduct C₁ C₂ ≤ maxTensorProduct C₁ C₂ := by
   exact Submodule.span_le.mpr (tmul_subset_maxTensorProduct C₁ C₂)
```
```suggestion
  exact Submodule.span_le.mpr (tmul_subset_maxTensorProduct C₁ C₂)

```

YaelDillies commented on Mathlib/Geometry/Convex/Cone/TensorProduct.lean:
```diff
@@ -0,0 +1,105 @@
+
+/-- Characterization of the maximal tensor product: `z` lies in
+`maxTensorProduct C₁ C₂` iff all pairings with elementary dual tensors are nonnegative. -/
+theorem mem_maxTensorProduct_iff {C₁ : PointedCone R G} {C₂ : PointedCone R H} {z : G ⊗[R] H} :
```
I think this lemma should be simp, akin to `mem_dual`
```suggestion
@[simp]
theorem mem_maxTensorProduct_iff {C₁ : PointedCone R G} {C₂ : PointedCone R H} {z : G ⊗[R] H} :
```

YaelDillies commented on Mathlib/Geometry/Convex/Cone/TensorProduct.lean:
```diff
@@ -0,0 +1,105 @@
+  · intro h
+    simp only [maxTensorProduct, mem_dual, minTensorProduct, dual_span]
+    rintro x ⟨φ, hφ, ψ, hψ, rfl⟩
+    exact h φ hφ ψ hψ
```
What stops the following proof to work?
```suggestion
  simp [maxTensorProduct, minTensorProduct]
```

bjornsolheim committed:
Rewriting and shortening proofs

bjornsolheim committed:
Added theorem tmul_subset_minTensorProduct

YaelDillies commented on Mathlib/Geometry/Convex/Cone/TensorProduct.lean:
```diff
@@ -44,41 +44,40 @@ We define the minimal and maximal tensor products of two pointed cones:
-      (.dual (Module.dualPairing K G).flip C₁.carrier)
-      (.dual (Module.dualPairing K H).flip C₂.carrier)).carrier
+      (.dual (Module.dualPairing R G).flip C₁.carrier)
+      (.dual (Module.dualPairing R H).flip C₂.carrier)).carrier
```
It's better to use the explicit type ascription

YaelDillies commented on Mathlib/Geometry/Convex/Cone/TensorProduct.lean:
```diff
@@ -0,0 +1,102 @@
+-/
+import Mathlib.Geometry.Convex.Cone.Dual
+import Mathlib.LinearAlgebra.TensorProduct.Basic
+import Mathlib.LinearAlgebra.Dual.Lemmas
```
```suggestion
import Mathlib.LinearAlgebra.Dual.Lemmas
import Mathlib.LinearAlgebra.TensorProduct.Basic
```

YaelDillies reviewed (approved):
Thanks!

maintainer delegate

bjornsolheim committed:
Reorganized import order. Opened Module and removed redundant namespace prefixes.

bjornsolheim commented:
Thanks. Thank you for a good process.

ocfnash reviewed (approved):
Thanks!

One suggestion about the doc string (feel free to edit my words) but otherwise LGTM.

bors d+

bjornsolheim committed:
The introductory part of the module doc string was rewritten for clarity and readability. A minor change in the doc string for minTensorProduct.

bjornsolheim committed:
minor rephrasing

ocfnash commented:
Thanks!

bors merge
